### PR TITLE
Computation of J.

### PR DESCRIPTION
**Pull request by favonia**
#627. I worried too much---this is straightforward (in the sense that almost every step is guided by the typing requirement and no creativity is involved, not that the whole expression is as succinct as in CCHM).

### CODE REVIEW

File: example/J.prl [hunk 1]
```diff
@@ -14,13 +22,33 @@ Thm J(#l:lvl) : [
...
+   [ty : (U #l kan)]
+   [a : ty]
+   [fam : (-> [x : ty] (path [_] ty a x) (U #l kan))]
+   [d : ($ fam a (abs [_] a))]
+   (path [_] ($ fam a (abs [_] a))
+    ($ (J #l) ty a fam d a (abs [_] a))
+    d))
+] by [
+  lam ty a fam d =>
+    `(abs [k]
```
Comment (jonsterling) [inline example/J.prl]: Does it work to use `abs k =>` here?

Comment (favonia) [reply example/J.prl]: Sure.

**Comment by favonia** [comment]
@mortberg Thanks for the review. Again, personally I do not feel any conceptual complexity because most steps are forced, but maybe you were saying the length of program is longer than the proof in CCHM? I agree it is longer. About the alternative J---the proof should work because it has nothing to do with the computation of connections. The faces of connections, not the computation of them, are the things that matter.

**Comment by mortberg** [comment]
About J defined using connections see my comment on the issue. I would be happy to be proved wrong! :)

By more complicated I mean that the length is a lot longer yes. My experience when formalizing in cubicaltt is that this kind of complications due to lack of regularity gets very annoying and tedious. This is really the simplest thing one should prove about J, so when doing much more complicated things using it one might have much bigger troubles because of things not disappearing by computation. I think you will see what I mean if you formalize pi4(s3). At some point I had to spend a few hundred lines just to get rid of things that should have disappeared by regularity... My conclusion was that one should never use J in cubicaltt unless one really has to.

**Comment by favonia** [comment]
Unfortunately, as you pointed out in the issue #627, it still needs extra fixing.

I totally agree with the comment about J. I believe this J is to satisfy the followers of intensional type theory and/or the book HoTT, not that it itself is particularly important to our development of other theorems. Seriously, just do the composition! The "true" J in my opinion lies in the different type that is inductively defined by refl, which will be there once Part IV is implemented in RedPRL. For that "true" J, things will compute definitionally.

**Pull request merged**